pkg/controller/dns/letsencrypt.go

```diff
@@ -0,0 +1,349 @@
+package dns
+
+import (
+	"context"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/sha1"
+	"crypto/x509"
+	"encoding/hex"
+	"encoding/json"
+	"encoding/pem"
+	"fmt"
+	"strings"
+	"time"
+
+	apiv1 "github.com/acorn-io/acorn/pkg/apis/api.acorn.io/v1"
+	"github.com/acorn-io/acorn/pkg/labels"
+	"github.com/acorn-io/acorn/pkg/system"
+	"github.com/acorn-io/baaah/pkg/router"
+	"github.com/go-acme/lego/v4/certcrypto"
+	"github.com/go-acme/lego/v4/certificate"
+	"github.com/go-acme/lego/v4/lego"
+	"github.com/go-acme/lego/v4/registration"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	LetsEncryptURLStaging    = "https://acme-staging-v02.api.letsencrypt.org/directory"
+	LetsEncryptURLProduction = "https://acme-v02.api.letsencrypt.org/directory"
+)
+
+type LEUser struct {
+	Name         string
+	Email        string
+	Registration *registration.Resource
+	Key          crypto.PrivateKey
+	URL          string
+}
+
+func FromSecret(secret *corev1.Secret) (*LEUser, error) {
+	block, _ := pem.Decode(secret.Data["privateKey"])
+	x509Encoded := block.Bytes
+	privateKey, err := x509.ParseECPrivateKey(x509Encoded)
+	if err != nil {
+		return nil, err
+	}
+
+	var reg registration.Resource
+	if err := json.Unmarshal(secret.Data["registration"], &reg); err != nil {
+		return nil, err
+	}
+
+	return &LEUser{
+		Name:         string(secret.Data["name"]),
+		Email:        string(secret.Data["email"]),
+		Registration: &reg,
+		Key:          privateKey,
+		URL:          string(secret.Data["url"]),
+	}, nil
+}
+
+func (u *LEUser) GetEmail() string {
+	return u.Email
+}
+func (u LEUser) GetRegistration() *registration.Resource {
+	return u.Registration
+}
+func (u *LEUser) GetPrivateKey() crypto.PrivateKey {
+	return u.Key
+}
+
+func (u *LEUser) Register() error {
+	if u.Name == "" || u.Email == "" {
+		return fmt.Errorf("not registering LE User: missing name or email")
+	}
+
+	if u.URL == "" {
+		return fmt.Errorf("not registering LE User: missing URL")
+	}
+
+	if u.Key == nil {
+		key, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+		if err != nil {
+			return err
+		}
+		u.Key = key
+	}
+
+	conf := lego.NewConfig(u)
+
+	conf.CADirURL = u.URL
+
+	client, err := lego.NewClient(conf)
+	if err != nil {
+		return err
+	}
+
+	reg, err := client.Registration.Register(registration.RegisterOptions{TermsOfServiceAgreed: true})
```

Comment:
We probably need to prompt the user to agree to LE's TOS.

I am thinking that if you say you are doing production LE, you can either pass the flag `--lets-encrypt-tos-agree=true` or if you don't do that, the CLI blocks and prints out a message like:

```
You are choosing to use Let's Encrypt production environment for TLS certificates. To do so, you must agree to their Terms of Serivice: https://community.letsencrypt.org/tos

Do you agree to Let's Encrypt TOS? y/n
```

Thoughts?

Reply:
Makes total sense, done!